File: zero-server/src/main/kotlin/com/zeroday/model/Notification.kt
package com.zeroday.model

import kotlinx.serialization.Serializable

/**
 * A persistent in-game message for the player. Notifications are appended
 * to the player's inbox (newest last) and capped to a fixed size so the
 * payload stays bounded. The client is expected to display the unread
 * ones in a notification center.
 */
@Serializable
data class Notification(
    val id: String = java.util.UUID.randomUUID().toString(),
    val type: NotificationType,
    val title: String,
    val message: String,
    val createdAt: Long = System.currentTimeMillis(),
    var read: Boolean = false,
    /** Optional structured payload so the client can deep-link. */
    val data: Map<String, String> = emptyMap()
)

enum class NotificationType {
    ACHIEVEMENT_UNLOCKED,
    CHALLENGE_COMPLETED,
    CHALLENGE_EXPIRED,
    STORYLINE_AVAILABLE,
    FACTION_INVITED,
    LEVEL_UP,
    WORLD_EVENT_STARTED,
    WORLD_EVENT_JOINABLE,
    SYSTEM_MESSAGE,
    REWARD_GRANTED,
    DAILY_LOGIN
}

File: zero-server-csharp/src/ZeroDayMMO.Shared/Models/Notification.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ZeroDayMMO.Shared.Models;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum NotificationType
{
    [JsonPropertyName("achievement_unlocked")]
    ACHIEVEMENT_UNLOCKED,
    [JsonPropertyName("challenge_completed")]
    CHALLENGE_COMPLETED,
    [JsonPropertyName("challenge_expired")]
    CHALLENGE_EXPIRED,
    [JsonPropertyName("storyline_available")]
    STORYLINE_AVAILABLE,
    [JsonPropertyName("faction_invited")]
    FACTION_INVITED,
    [JsonPropertyName("level_up")]
    LEVEL_UP,
    [JsonPropertyName("world_event_started")]
    WORLD_EVENT_STARTED,
    [JsonPropertyName("world_event_joinable")]
    WORLD_EVENT_JOINABLE,
    [JsonPropertyName("system_message")]
    SYSTEM_MESSAGE,
    [JsonPropertyName("reward_granted")]
    REWARD_GRANTED,
    [JsonPropertyName("daily_login")]
    DAILY_LOGIN
}

public class Notification
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = Guid.NewGuid().ToString();

    [JsonPropertyName("type")]
    public NotificationType Type { get; set; }

    [JsonPropertyName("title")]
    public string Title { get; set; } = "";

    [JsonPropertyName("message")]
    public string Message { get; set; } = "";

    [JsonPropertyName("created_at")]
    public long CreatedAt { get; set; } = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    [JsonPropertyName("read")]
    public bool Read { get; set; } = false;

    [JsonPropertyName("data")]
    public Dictionary<string, string> Data { get; set; } = new();
}
